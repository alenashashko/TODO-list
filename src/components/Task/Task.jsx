import React, {useState, useCallback} from "react";

import TaskViewMode from "./TaskViewMode/TaskViewMode";
import TaskEditMode from "./TaskEditMode/TaskEditMode";

export const Task = (props) => {
  const {task} = props;

  const [isViewMode, setIsViewMode] = useState(true);

  const handleEditMode = useCallback(() => {
    setIsViewMode(false);
  }, []);

  const handleViewMode = useCallback(() => {
    setIsViewMode(true);
  }, []);

  return isViewMode
    ? <TaskViewMode task={task} onEditClick={handleEditMode} />
    : <TaskEditMode task={task} onFinish={handleViewMode} />
};
