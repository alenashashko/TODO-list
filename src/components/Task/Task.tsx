import { FC, useState, useCallback } from "react";

import { TaskViewMode } from "./TaskViewMode/TaskViewMode";
import { TaskEditMode } from "./TaskEditMode/TaskEditMode";

export interface Task {
  id: string;
  date: number;
  text: string;
}

interface TaskProps {
  task: Task;
  initialIsViewMode?: boolean;
}

export const useIsViewMode = (initialValue = true) => {
  const [isViewMode, setIsViewMode] = useState<boolean>(initialValue);

  const handleEditMode = useCallback(() => {
    setIsViewMode(false);
  }, []);

  const handleViewMode = useCallback(() => {
    setIsViewMode(true);
  }, []);

  return [isViewMode, handleViewMode, handleEditMode] as const;
};

export const Task: FC<TaskProps> = ({ task, initialIsViewMode = true }) => {
  const [isViewMode, handleViewMode, handleEditMode] = useIsViewMode(
    initialIsViewMode
  );

  return isViewMode ? (
    <TaskViewMode task={task} onEditClick={handleEditMode} />
  ) : (
    <TaskEditMode task={task} onFinish={handleViewMode} />
  );
};
