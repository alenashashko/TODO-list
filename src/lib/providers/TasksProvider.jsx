import React, {useContext, useState, useCallback, useMemo} from "react";
import {nanoid} from "nanoid";

const TasksContext = React.createContext({
  tasks: []
});

export const useTasksInfo = () => {
  const value = useContext(TasksContext);

  if (value === undefined) {
    throw new Error("you should use useTasksInfo inside of TasksProvider");
  }

  return value;
};

export const TasksProvider = (props) => {
  const {children} = props;

  const [tasks, setTasks] = useState([]);

  const createTask = useCallback((text) => {
    setTasks([...tasks, {
      id: nanoid(),
      date: Date.now(),
      text
    }])
  }, [tasks]);

  const changeTask = useCallback((taskId, newText) => {
    setTasks(tasks.map((task) => {
      return task.id !== taskId
        ? task
        : {
          ...task,
          date: Date.now(),
          text: newText
        }
    }))
  }, [tasks]);

  const deleteTask = useCallback((taskId) => {
    setTasks(tasks.filter((task) => task.id !== taskId));
  }, [tasks]);

  const value = useMemo(() => ({
    tasks,
    createTask,
    changeTask,
    deleteTask
  }), [tasks, createTask, changeTask, deleteTask]);

  return <TasksContext.Provider value={value}>{children}</TasksContext.Provider>;
};
