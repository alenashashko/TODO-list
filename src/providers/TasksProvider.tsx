import {FC, createContext, useContext, useState, useCallback, useMemo} from "react";
import {nanoid} from "nanoid";
import { Task } from "components/Task/Task";

const TasksContext = createContext<{
  tasks: Task[];
  createTask: (text: string) => void;
  changeTask: (taskId: string, newText: string) => void;
  deleteTask: (taskId: string) => void;
}>({
  tasks: [],
  createTask: () => {},
  changeTask: () => {},
  deleteTask: () => {},
});

export const useTasksInfo = () => {
  const value = useContext(TasksContext);

  if (value === undefined) {
    throw new Error("you should use useTasksInfo inside of TasksProvider");
  }

  return value;
};

export const TasksProvider: FC = (props) => {
  const {children} = props;

  const [tasks, setTasks] = useState<Task[]>([]);

  const createTask = useCallback((text: string) => {
    setTasks([...tasks, {
      id: nanoid(),
      date: Date.now(),
      text
    }])
  }, [tasks]);

  const changeTask = useCallback((taskId: string, newText: string) => {
    setTasks(tasks.map((task) => {
      return task.id !== taskId
        ? task
        : {
          ...task,
          date: Date.now(),
          text: newText
        }
    }))
  }, [tasks]);

  const deleteTask = useCallback((taskId: string) => {
    setTasks(tasks.filter((task) => task.id !== taskId));
  }, [tasks]);

  const value = useMemo(() => ({
    tasks,
    createTask,
    changeTask,
    deleteTask
  }), [tasks, createTask, changeTask, deleteTask]);

  return <TasksContext.Provider value={value}>{children}</TasksContext.Provider>;
};
