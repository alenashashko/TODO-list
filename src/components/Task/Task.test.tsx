import { render, cleanup, act } from "@testing-library/react";
import { renderHook } from "@testing-library/react-hooks";
import { TasksProvider } from "src/providers/TasksProvider";
import { taskMock } from "src/tests/mocks/tasks";

import { Task, useIsViewMode } from "./Task";

afterEach(cleanup);

describe("Task Component", () => {
  it("renders task in view mode", () => {
    const { asFragment } = render(
      <TasksProvider>
        <Task task={taskMock} initialIsViewMode={true} />
      </TasksProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders task in edit mode", () => {
    const { asFragment } = render(
      <TasksProvider>
        <Task task={taskMock} initialIsViewMode={false} />
      </TasksProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});

describe("useIsViewMode hook", () => {
  it("should be true by default", () => {
    const { result } = renderHook(() => useIsViewMode());

    expect(result.current[0]).toBe(true);
  });

  it("should be false if hook argument is false", () => {
    const { result } = renderHook(() => useIsViewMode(false));

    expect(result.current[0]).toBe(false);
  });

  it("should be true if handleViewMode is called", () => {
    const { result } = renderHook(() => useIsViewMode(false));

    act(() => {
      const handleViewMode = result.current[1];

      handleViewMode();
    });

    expect(result.current[0]).toBe(true);
  });

  it("should be false if handleEditMode is called", () => {
    const { result } = renderHook(() => useIsViewMode(true));

    act(() => {
      const handleEditMode = result.current[2];

      handleEditMode();
    });

    expect(result.current[0]).toBe(false);
  });
});
