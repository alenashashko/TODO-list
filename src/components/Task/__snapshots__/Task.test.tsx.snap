// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Component renders task in edit mode 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-task-6 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardContent-root makeStyles-content-7"
    >
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-text-8"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-multiline MuiInput-multiline"
        >
          <textarea
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline"
            name="task"
            placeholder="Start typing your text here..."
            rows="1"
            style="height: 0px; overflow: hidden;"
          >
            my task
          </textarea>
          <textarea
            aria-hidden="true"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline"
            readonly=""
            style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 100%;"
            tabindex="-1"
          />
        </div>
      </div>
    </div>
    <div
      class="MuiCardActions-root makeStyles-footer-9 MuiCardActions-spacing"
    >
      <div
        class="makeStyles-footerContainer-10"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
        >
          03/09/2021 3:50 PM
        </p>
        <div
          class="makeStyles-buttons-11"
        >
          <span
            class=""
            title="Save"
          >
            <button
              aria-label="save changes"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled MuiIconButton-sizeSmall Mui-disabled"
              data-testid="save"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
                  />
                </svg>
              </span>
            </button>
          </span>
          <button
            aria-label="cancel changes"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeSmall"
            data-testid="cancel"
            tabindex="0"
            title="Cancel"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Task Component renders task in view mode 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-task-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardContent-root"
    >
      <p
        class="MuiTypography-root makeStyles-text-2 MuiTypography-body1"
      >
        my task
      </p>
    </div>
    <div
      class="MuiCardActions-root makeStyles-footer-3 MuiCardActions-spacing"
    >
      <div
        class="makeStyles-footerContainer-4"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
        >
          03/09/2021 3:50 PM
        </p>
        <div
          class="makeStyles-buttons-5"
        >
          <button
            aria-label="edit task"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall"
            data-testid="edit"
            tabindex="0"
            title="Edit"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="delete task"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeSmall"
            data-testid="delete"
            tabindex="0"
            title="Delete"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
